dev/archery/archery/integration/datagen.py

```diff
@@ -221,6 +221,30 @@ def _get_type(self):
             ('unit', 'DAY' if self.unit == self.DAY else 'MILLISECOND')
         ])
 
+    def generate_range(self, size, lower, upper, name=None,
+                       include_extremes=False):
+        if self.unit == self.DAY:
+            return super().generate_range(size, lower, upper, name)
+
+        full_day_millis = 1000 * 60 * 60 * 24
+        lower //= full_day_millis
```

Comment:
Python appears to round towards -Inf so a lower negative bound that is not exactly even will get rounded to an out-of-bound value